src/daemon/rpm-ostreed.service.in

```diff
@@ -20,6 +20,11 @@ ProtectHome=true
 # entry here ensures that the skopeo process we fork won't interact with
 # application containers.
 InaccessiblePaths=-/var/lib/containers
+# Some popular akmods, when invoked through akmods-ostree-post, will want to
+# open more files than we would normally have access to. This is sometimes a
+# function of the number of logical CPU cores, and so we are generous in our
+# request of systemd.
+LimitNOFILE=65535
```

Comment:
What host operating system are you using?  Quite a while ago, systemd bumped the default `LimitNoFile`.  

```
[root@cosa-devsh ~]# rpm-ostree status
State: idle
Deployments:
* fedora:fedora/x86_64/coreos/testing-devel
                   Version: 36.20220715.dev.1 (2022-07-15T11:58:03Z)
                    Commit: 4d99e9a5172b7734df6bb915d6dce80499a756b4d6756e69c3257945c6ae9594
              GPGSignature: (unsigned)
[root@cosa-devsh ~]# systemctl --version
systemd 250 (v250.7-1.fc36)
+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
[root@cosa-devsh ~]# systemctl show rpm-ostreed | grep -ie 'Limit.*File'
LimitNOFILE=524288
LimitNOFILESoft=1024
[root@cosa-devsh ~]# 
```

So in fact you'd be *lowering* the (hard) limit today, not raising it.

Reply:
I'll drop a full analysis later this week for posterity when I have time to reproduce in a VM, but the crux is that the soft limit is the problem, not the hard limit. Indeed, **this change does lower the hard limit**, but as far as I can tell it's impossible to raise _only_ the soft limit — In https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Process%20Properties I can only see how to either 1) set both properties, with `LimitNOFILE=<value>`, or set hard and soft specifically, with `LimitNOFILE=<soft>:<hard>`. In my experimentation, something like `LimitNOFILE=<soft>:` doesn't apply, and usage of the old `LimitNOFILESoft` key is ignored with a warning in `systemctl status rpm-ostreed.service`.

I'll leave this conversation unresolved until I have a moment to drop the full analysis for others to follow.

Reply:
Systemd defaults to a low soft limit for a reason; it's basically for backwards compatibility with applications still using `select()` (instead of e.g. `epoll()` like modern apps should use).  Raising the soft limit can cause those applications using `select()` to run out of stack.

I think basically what instead needs to change is the build script involved here should try to raise the soft limit to match the hard limit, via e.g. `ulimit -S -n hard`.